asyncpg/pool.py

```diff
@@ -605,15 +605,29 @@ async def _acquire_impl():
                 ch._timeout = timeout
                 return proxy
 
+        def _release_leaked_connection(fut):
+            if not fut.cancelled():
+                asyncio.ensure_future(self.release(fut.result()))
+
         if self._closing:
             raise exceptions.InterfaceError('pool is closing')
         self._check_init()
 
+        acquire_fut = asyncio.ensure_future(_acquire_impl())
         if timeout is None:
-            return await _acquire_impl()
+            return await acquire_fut
```

Comment:
The leak might be possible even when `timeout` is `None`.

Reply:
So I *think* there shouldn't be the leak here, as a future will go from `pending` to `done` or `cancelled` (i.e can't go from `done` to `cancelled`), but this is getting into the internals of asyncio, so there may be some gotchas. This overall issue arises due to the `await` within `wait_for` after the future has resolved, which is where the cancellation is able to sneak in.

Either way, we could just ditch the `if timeout is None` conditional and always use the `wait_for` branch (with the callback), as passing `timeout = None` has the desired behaviour.

Reply:
Updated version pushed with the above change, which will use wait_for regardless and always catch the cancellation.